```vast-cc``` fails to emits mlir correctly

When I tried to compile r1.c to clang ast and emits mlir in the high-level dialect by executing ```vast-cc --from-source   --no-warn  --instcombine-code-sinking     --asm-show-inst -o ./r1.mlir r1.c```.
r1.c:
```console
#define NO_LONGLONG
#include "csmith.h"

static long __undefined;

static uint8_t  func_1(void)
{ /* block id: 0 */
    int32_t l_4 = 0;
    int32_t l_5[7] = {0,0,0,0,0,0,0};
    int i;
    l_5[6] = (safe_add_func_uint32_t_u_u(4294967293U, l_4));
    return l_5[6];
}

/* ---------------------------------------- */
int main (int argc, char* argv[])
{
    int print_hash_value = 0;
    if (argc == 2 && strcmp(argv[1], "1") == 0) print_hash_value = 1;
    platform_main_begin();
    crc32_gentab();
    func_1();
    platform_main_end(crc32_context ^ 0xFFFFFFFFUL, print_hash_value);
    return 0;
}

```
csmith.h:
```console
#ifndef RANDOM_RUNTIME_H
#define RANDOM_RUNTIME_H

#ifdef CSMITH_MINIMAL
#include "csmith_minimal.h"
#else

#include <string.h>
#include <float.h>
#include <math.h>

#define __STDC_LIMIT_MACROS
#include "random_inc.h"

static uint32_t crc32_tab[256];
static uint32_t crc32_context = 0xFFFFFFFFUL;

static void 
crc32_gentab (void)
{
	uint32_t crc;
	const uint32_t poly = 0xEDB88320UL;
	int i, j;
	
	for (i = 0; i < 256; i++) {
		crc = i;
		for (j = 8; j > 0; j--) {
			if (crc & 1) {
				crc = (crc >> 1) ^ poly;
			} else {
				crc >>= 1;
			}
		}
		crc32_tab[i] = crc;
	}
}

static void 
crc32_byte (uint8_t b) {
	crc32_context = 
		((crc32_context >> 8) & 0x00FFFFFF) ^ 
		crc32_tab[(crc32_context ^ b) & 0xFF];
}

#if defined(__SPLAT__) || defined(NO_LONGLONG)
static void 
crc32_8bytes (uint32_t val)
{
	crc32_byte ((val>>0) & 0xff);
	crc32_byte ((val>>8) & 0xff);
	crc32_byte ((val>>16) & 0xff);
	crc32_byte ((val>>24) & 0xff);
}

static void 
transparent_crc (uint32_t val, char* vname, int flag)
{
	crc32_8bytes(val);
	if (flag) {
  		printf("...checksum after hashing %s : %X\n", vname, crc32_context ^ 0xFFFFFFFFU);
	}
}
#else
static void 
crc32_8bytes (uint64_t val)
{
	crc32_byte ((val>>0) & 0xff);
	crc32_byte ((val>>8) & 0xff);
	crc32_byte ((val>>16) & 0xff);
	crc32_byte ((val>>24) & 0xff);
	crc32_byte ((val>>32) & 0xff);
	crc32_byte ((val>>40) & 0xff);
	crc32_byte ((val>>48) & 0xff);
	crc32_byte ((val>>56) & 0xff);
}

static void 
transparent_crc (uint64_t val, char* vname, int flag)
{
	crc32_8bytes(val);
	if (flag) {
  		printf("...checksum after hashing %s : %lX\n", vname, crc32_context ^ 0xFFFFFFFFUL);
	}
}

#endif

static void 
transparent_crc_bytes (char *ptr, int nbytes, char* vname, int flag)
{
    int i;
    for (i=0; i<nbytes; i++) {
        crc32_byte(ptr[i]);
    }
	if (flag) {
  		printf("...checksum after hashing %s : %lX\n", vname, crc32_context ^ 0xFFFFFFFFUL);
	}
}
#endif
#endif /* RANDOM_RUNTIME_H */
```
It returned with:
```console
In file included from input.cc:13:
In file included from ./csmith.h:45:
In file included from ./random_inc.h:100:
./safe_math.h:1201:36: warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
    ((sf1 <= INT32_MIN) || (sf1 >= INT32_MAX)) ?
                                ~~ ^~~~~~~~~
/usr/include/stdint.h:123:22: note: expanded from macro 'INT32_MAX'
# define INT32_MAX              (2147483647)
                                 ^~~~~~~~~~
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: vast-cc --from-source --no-warn --instcombine-code-sinking --asm-show-inst -o ./r1.mlir r1.c
 #0 0x00007f853410e3b1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-15/lib/libLLVM-15.so.1+0xf043b1)
 #1 0x00007f853410c0fe llvm::sys::RunSignalHandlers() (/usr/lib/llvm-15/lib/libLLVM-15.so.1+0xf020fe)
 #2 0x00007f853410e8d6 (/usr/lib/llvm-15/lib/libLLVM-15.so.1+0xf048d6)
 #3 0x00007f8532cee520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00005563f16b902d mlir::Operation* vast::cg::CodeGenStmtVisitorMixin<vast::cg::CodeGenVisitor<vast::cg::DefaultCodeGenVisitorConfig, vast::cg::DefaultMetaGenerator> >::VisitCast<vast::hl::ImplicitCastOp>(clang::CastExpr const*) FromSource.cpp:0:0
 #5 0x00005563f16b328c vast::cg::CodeGenStmtVisitorMixin<vast::cg::CodeGenVisitor<vast::cg::DefaultCodeGenVisitorConfig, vast::cg::DefaultMetaGenerator> >::VisitReturnStmt(clang::ReturnStmt const*) FromSource.cpp:0:0
 #6 0x00005563f16b2e1d vast::cg::CodeGenStmtVisitorMixin<vast::cg::CodeGenVisitor<vast::cg::DefaultCodeGenVisitorConfig, vast::cg::DefaultMetaGenerator> >::VisitCompoundStmt(clang::CompoundStmt const*) FromSource.cpp:0:0
 #7 0x00005563f16b2445 auto vast::cg::CodeGenDeclVisitorMixin<vast::cg::CodeGenVisitor<vast::cg::DefaultCodeGenVisitorConfig, vast::cg::DefaultMetaGenerator> >::VisitFunctionDecl(clang::FunctionDecl const*)::'lambda1'(auto)::operator()<vast::hl::FuncOp>(auto) const FromSource.cpp:0:0
 #8 0x00005563f16ae7f8 vast::cg::CodeGenDeclVisitorMixin<vast::cg::CodeGenVisitor<vast::cg::DefaultCodeGenVisitorConfig, vast::cg::DefaultMetaGenerator> >::VisitFunctionDecl(clang::FunctionDecl const*) FromSource.cpp:0:0
 #9 0x00005563f16adec6 vast::cg::CodeGenBase<vast::cg::CodeGenVisitor<vast::cg::DefaultCodeGenVisitorConfig, vast::cg::DefaultMetaGenerator> >::process_root_decl(void*, clang::Decl const*) FromSource.cpp:0:0
#10 0x00005563f26a71fa clang::ASTUnit::visitLocalTopLevelDecls(void*, bool (*)(void*, clang::Decl const*)) (/data/xcy/vast/builds/ninja-multi-default/tools/vast-cc/Release/vast-cc+0x1eb91fa)
#11 0x00005563f168b1a0 std::_Function_handler<mlir::OwningOpRef<mlir::ModuleOp> (llvm::SourceMgr&, mlir::MLIRContext*), vast::hl::registerFromSourceParser()::$_1>::_M_invoke(std::_Any_data const&, llvm::SourceMgr&, mlir::MLIRContext*&&) FromSource.cpp:0:0
#12 0x00005563f265881a std::_Function_handler<mlir::LogicalResult (llvm::SourceMgr&, llvm::raw_ostream&, mlir::MLIRContext*), registerTranslateToMLIRFunction(llvm::StringRef, std::function<mlir::OwningOpRef<mlir::ModuleOp> (llvm::SourceMgr&, mlir::MLIRContext*)> const&)::$_0>::_M_invoke(std::_Any_data const&, llvm::SourceMgr&, llvm::raw_ostream&, mlir::MLIRContext*&&) HighLevelLinkage.cpp:0:0
#13 0x00005563f2657605 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&)>::callback_fn<mlir::mlirTranslateMain(int, char**, llvm::StringRef)::$_0>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&) HighLevelLinkage.cpp:0:0
#14 0x00005563f2657a61 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/data/xcy/vast/builds/ninja-multi-default/tools/vast-cc/Release/vast-cc+0x1e69a61)
#15 0x00005563f2655eb6 mlir::mlirTranslateMain(int, char**, llvm::StringRef) (/data/xcy/vast/builds/ninja-multi-default/tools/vast-cc/Release/vast-cc+0x1e67eb6)
#16 0x00005563f168a042 main (/data/xcy/vast/builds/ninja-multi-default/tools/vast-cc/Release/vast-cc+0xe9c042)
#17 0x00007f8532cd5d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#18 0x00007f8532cd5e40 call_init ./csu/../csu/libc-start.c:128:20
#19 0x00007f8532cd5e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#20 0x00005563f1689ea5 _start (/data/xcy/vast/builds/ninja-multi-default/tools/vast-cc/Release/vast-cc+0xe9bea5)
Segmentation fault (core dumped)
```
However, I can compiles r1.c with gcc correctly. Could anyone know what should I do next?

--- a/include/vast/Dialect/HighLevel/HighLevelOps.td
+++ b/include/vast/Dialect/HighLevel/HighLevelOps.td
@@ -879,7 +879,6 @@ def Predicate : PredicateList< "Predicate", "comparison predicate", [
   ICmpPredicateUGT, ICmpPredicateUGE
 ] >;
 
-
 def CmpOp
   : HighLevel_Op< "cmp", [NoSideEffect] >
   , Arguments<(ins Predicate:$predicate, AnyType:$lhs, AnyType:$rhs)>
@@ -890,6 +889,48 @@ def CmpOp
 
   let assemblyFormat = "$predicate $lhs `,` $rhs  attr-dict `:` type(operands) `->` type($result)";
 }
+
+class FPredicateAttr< string name, int val > : I64EnumAttrCase< name, val > {}
+
+class FPredicateList< string name, string summary, list<FPredicateAttr> cases >
+  : I64EnumAttr< name, summary, cases > {}
+
+def FPredicateFalse : FPredicateAttr<"ffalse", 0>;
+def FPredicateOEQ   : FPredicateAttr<"oeq",    1>;
+def FPredicateOGT   : FPredicateAttr<"ogt",    2>;
+def FPredicateOGE   : FPredicateAttr<"oge",    3>;
+def FPredicateOLT   : FPredicateAttr<"olt",    4>;
+def FPredicateOLE   : FPredicateAttr<"ole",    5>;
+def FPredicateONE   : FPredicateAttr<"one",    6>;
+def FPredicateORD   : FPredicateAttr<"ord",    7>;
+def FPredicateUNO   : FPredicateAttr<"uno",    8>;
+def FPredicateUEQ   : FPredicateAttr<"ueq",    9>;
+def FPredicateUGT   : FPredicateAttr<"ugt",   10>;
+def FPredicateUGE   : FPredicateAttr<"uge",   11>;
+def FPredicateULT   : FPredicateAttr<"ult",   12>;
+def FPredicateULE   : FPredicateAttr<"ule",   13>;
+def FPredicateUNE   : FPredicateAttr<"une",   14>;
+def FPredicateTrue  : FPredicateAttr<"ftrue", 15>;
+
+let cppNamespace = "::vast::hl" in
+def FPredicate : FPredicateList< "FPredicate", "floating point comparison predicate", [
+  FPredicateFalse, FPredicateOEQ, FPredicateOGT, FPredicateOGE,
+  FPredicateOLT, FPredicateOLE, FPredicateONE, FPredicateORD,
+  FPredicateUNO, FPredicateUEQ, FPredicateUGT, FPredicateUGE,
+  FPredicateULT, FPredicateULE, FPredicateUNE, FPredicateTrue
+] >;
+
+def FCmpOp
+  : HighLevel_Op< "fcmp", [NoSideEffect] >
+  , Arguments<(ins FPredicate:$predicate, FloatLikeType:$lhs, FloatLikeType:$rhs)>
+  , Results<(outs IntOrBoolType:$result)>
+{
+  let summary = "VAST flaoting point comparison operation";
+  let description = [{ VAST flaoting point comparison operation }];
+
+  let assemblyFormat = "$predicate $lhs `,` $rhs  attr-dict `:` type(operands) `->` type($result)";
+}
+
 
 
 class UnInplaceOp< string mnemonic, list< Trait > traits = [] >

--- a/include/vast/Dialect/HighLevel/HighLevelOps.td
+++ b/include/vast/Dialect/HighLevel/HighLevelOps.td
@@ -1089,4 +1089,21 @@ def AlignOfExprOp : ExprTraitOp< "alignof.expr" > {
   let description = [{ VAST expr alignof operator }];
 }
 
+def StmtExprOp
+  : HighLevel_Op< "stmt.expr", [RecursiveSideEffects, SingleBlock] >
+  , Results<(outs AnyType:$result)>
+{
+  let summary = "VAST statement expression";
+  let description = [{ VAST statement expression }];
+
+  let regions = (region SizedRegion<1>:$substmt);
+
+  let skipDefaultBuilders = 1;
+  let builders = [ OpBuilder<(ins "Type":$rty, "std::unique_ptr< Region > &&":$region)> ];
+
+  let assemblyFormat = [{
+     attr-dict `:` type($result) $substmt
+  }];
+}
+
 #endif // VAST_DIALECT_HIGHLEVEL_IR_HIGHLEVELOPS

--- a/include/vast/Translation/CodeGenBuilder.hpp
+++ b/include/vast/Translation/CodeGenBuilder.hpp
@@ -172,14 +172,29 @@ namespace vast::cg {
             return reg;
         }
 
+        std::unique_ptr< Region > make_stmt_region(const clang::CompoundStmt *stmt) {
+            auto guard = insertion_guard();
+
+            auto reg = std::make_unique< Region >();
+            set_insertion_point_to_start( &reg->emplaceBlock() );
+            for (auto s : stmt->body()) {
+                visit(s);
+            }
+
+            return reg;
+        }
+
         using RegionAndType = std::pair< std::unique_ptr< Region >, Type >;
-        RegionAndType make_value_yield_region(const clang::Stmt *stmt) {
+
+        template< typename StmtType >
+        RegionAndType make_value_yield_region(const StmtType *stmt) {
             auto guard  = insertion_guard();
             auto reg    = make_stmt_region(stmt);
 
             auto &block = reg->back();
             set_insertion_point_to_end( &block );
             auto type = block.back().getResult(0).getType();
+            VAST_CHECK(type, "value region require last operation to be value");
             create< hl::ValueYieldOp >(meta_location(stmt), block.back().getResult(0));
 
             return { std::move(reg), type };

--- a/include/vast/Translation/CodeGenStmtVisitor.hpp
+++ b/include/vast/Translation/CodeGenStmtVisitor.hpp
@@ -832,7 +832,8 @@ namespace vast::cg {
         // Operation* VisitParenListExpr(const clang::ParenListExpr *expr)
         Operation* VisitStmtExpr(const clang::StmtExpr *expr) {
             auto loc = meta_location(expr);
-            auto [reg, rty] = make_value_yield_region(expr->getSubStmt());
+            auto sub = llvm::cast< clang::CompoundStmt >(expr->getSubStmt());
+            auto [reg, rty] = make_value_yield_region(sub);
             return make< hl::StmtExprOp >(loc, rty, std::move(reg));
         }
 

--- a/include/vast/Translation/CodeGenStmtVisitor.hpp
+++ b/include/vast/Translation/CodeGenStmtVisitor.hpp
@@ -117,8 +117,16 @@ namespace vast::cg {
             return make< hl::CmpOp >(meta_location(op), res, pred, lhs, rhs);
         }
 
-        template< hl::Predicate upred, hl::Predicate spred >
-        Operation* VisitICmp(const clang::BinaryOperator *op) {
+        template< hl::FPredicate pred >
+        Operation* VisitFCmp(const clang::BinaryOperator *op) {
+            auto lhs = visit(op->getLHS())->getResult(0);
+            auto rhs = visit(op->getRHS())->getResult(0);
+            auto res = visit(op->getType());
+            return make< hl::FCmpOp >(meta_location(op), res, pred, lhs, rhs);
+        }
+
+        template< hl::Predicate upred, hl::Predicate spred, hl::FPredicate fpred >
+        Operation* VisitCmp(const clang::BinaryOperator *op) {
             auto ty = op->getLHS()->getType();
             if (ty->isUnsignedIntegerType())
                 return VisitCmp< upred >(op);
@@ -126,6 +134,8 @@ namespace vast::cg {
                 return VisitCmp< upred >(op);
             if (ty->isIntegerType())
                 return VisitCmp< spred >(op);
+            if (ty->isFloatingType())
+                return VisitFCmp< fpred >(op);
             return nullptr;
         }
 
@@ -160,28 +170,31 @@ namespace vast::cg {
             return VisitBinOp< hl::BinShrOp >(op);
         }
 
+        using ipred = hl::Predicate;
+        using fpred = hl::FPredicate;
+
         Operation* VisitBinLT(const clang::BinaryOperator *op) {
-            return VisitICmp< hl::Predicate::ult, hl::Predicate::slt >(op);
+            return VisitCmp< ipred::ult, ipred::slt, fpred::olt >(op);
         }
 
         Operation* VisitBinGT(const clang::BinaryOperator *op) {
-            return VisitICmp< hl::Predicate::ugt, hl::Predicate::sgt >(op);
+            return VisitCmp< ipred::ugt, ipred::sgt, fpred::ogt >(op);
         }
 
         Operation* VisitBinLE(const clang::BinaryOperator *op) {
-            return VisitICmp< hl::Predicate::ule, hl::Predicate::sle >(op);
+            return VisitCmp< ipred::ule, ipred::sle, fpred::ole >(op);
         }
 
         Operation* VisitBinGE(const clang::BinaryOperator *op) {
-            return VisitICmp< hl::Predicate::uge, hl::Predicate::sge >(op);
+            return VisitCmp< ipred::uge, ipred::sge, fpred::oge >(op);
         }
 
         Operation* VisitBinEQ(const clang::BinaryOperator *op) {
-            return VisitCmp< hl::Predicate::eq >(op);
+            return VisitCmp< ipred::eq, ipred::eq, fpred::oeq >(op);
         }
 
         Operation* VisitBinNE(const clang::BinaryOperator *op) {
-            return VisitCmp< hl::Predicate::ne >(op);
+            return VisitCmp< ipred::ne, ipred::ne, fpred::une >(op);
         }
 
         Operation* VisitBinAnd(const clang::BinaryOperator *op) {

--- a/include/vast/Translation/CodeGenStmtVisitor.hpp
+++ b/include/vast/Translation/CodeGenStmtVisitor.hpp
@@ -825,12 +825,16 @@ namespace vast::cg {
         // Operation* VisitOverloadExpr(const clang::OverloadExpr *expr)
 
         Operation* VisitParenExpr(const clang::ParenExpr *expr) {
-            auto [region, type] = make_value_yield_region(expr->getSubExpr());
-            return make< hl::ExprOp >(meta_location(expr), type, std::move(region));
+            auto [reg, rty] = make_value_yield_region(expr->getSubExpr());
+            return make< hl::ExprOp >(meta_location(expr), rty, std::move(reg));
         }
 
         // Operation* VisitParenListExpr(const clang::ParenListExpr *expr)
-        // Operation* VisitStmtExpr(const clang::StmtExpr *expr)
+        Operation* VisitStmtExpr(const clang::StmtExpr *expr) {
+            auto loc = meta_location(expr);
+            auto [reg, rty] = make_value_yield_region(expr->getSubStmt());
+            return make< hl::StmtExprOp >(loc, rty, std::move(reg));
+        }
 
         template< typename Op >
         Operation* ExprTypeTrait(const clang::UnaryExprOrTypeTraitExpr *expr, auto rty, auto loc) {

--- a/include/vast/Util/Warnings.hpp
+++ b/include/vast/Util/Warnings.hpp
@@ -51,11 +51,15 @@ VAST_UNRELAX_WARNINGS
 
 namespace vast
 {
-    #define VAST_UNREACHABLE(fmt, ...) llvm_unreachable( llvm::formatv(fmt __VA_OPT__(,) __VA_ARGS__).str().c_str() );
+    #define VAST_REPORT(...) llvm::dbgs() << "[vast] " << llvm::formatv(__VA_ARGS__) << "\n";
+
+    #define VAST_UNREACHABLE(...) \
+      VAST_REPORT(__VA_ARGS__) \
+      llvm_unreachable(nullptr);
 
     #define VAST_UNIMPLEMENTED VAST_UNREACHABLE("not implemented: {0}", __PRETTY_FUNCTION__);
 
-    #define VAST_DEBUG(fmt, ...) LLVM_DEBUG(llvm::dbgs() << llvm::formatv(fmt, __VA_OPT__(,) __VA_ARGS__));
+    #define VAST_DEBUG(fmt, ...) LLVM_DEBUG(VAST_REPORT(__VA_ARGS__))
 
     #define VAST_CHECK(cond, fmt, ...) if (!(cond)) { VAST_UNREACHABLE(fmt __VA_OPT__(,) __VA_ARGS__); }
 

--- a/lib/vast/Dialect/HighLevel/HighLevelLinkage.cpp
+++ b/lib/vast/Dialect/HighLevel/HighLevelLinkage.cpp
@@ -140,11 +140,12 @@ namespace vast::hl {
             case GlobalLinkageKind::PrivateLinkage:
                 return Visibility::Private;
             case GlobalLinkageKind::ExternalLinkage:
+            case GlobalLinkageKind::AvailableExternallyLinkage:
             case GlobalLinkageKind::ExternalWeakLinkage:
             case GlobalLinkageKind::LinkOnceODRLinkage:
                 return Visibility::Public;
             default:
-                VAST_UNREACHABLE("unsupported linkage kind");
+                VAST_UNREACHABLE("unsupported linkage kind {0}", stringifyGlobalLinkageKind(linkage));
         }
 
         VAST_UNREACHABLE("missed linkage kind");

--- a/lib/vast/Dialect/HighLevel/HighLevelOps.cpp
+++ b/lib/vast/Dialect/HighLevel/HighLevelOps.cpp
@@ -187,6 +187,12 @@ namespace vast::hl
         build_expr_trait(bld, st, rty, expr);
     }
 
+    void StmtExprOp::build(Builder &bld, State &st, Type rty, std::unique_ptr< Region > &&region) {
+        InsertionGuard guard(bld);
+        st.addRegion(std::move(region));
+        st.addTypes(rty);
+    }
+
     void VarDeclOp::build(Builder &bld, State &st, Type type, llvm::StringRef name, BuilderCallback init, BuilderCallback alloc) {
         st.addAttribute("name", bld.getStringAttr(name));
         InsertionGuard guard(bld);

--- a/test/vast/Dialect/HighLevel/stmt-expr.c
+++ b/test/vast/Dialect/HighLevel/stmt-expr.c
@@ -1,0 +1,10 @@
+// RUN: vast-cc --ccopts -xc --from-source %s | FileCheck %s
+// RUN: vast-cc --ccopts -xc --from-source %s > %t && vast-opt %t | diff -B %t -
+
+int main() {
+    // CHECK: hl.stmt.expr : !hl.int
+    // CHECK:   hl.var "x"
+    // CHECK:   hl.value.yield
+    // CHECK:   hl.value.yield
+    int v = ({int x = 4; x;});
+}

